### ISSUE TITLE
Regression after stockfish-nnue-2020-08-30

### ISSUE THREAD

**Issue by nodchip** [nodchip/Stockfish#152]
In #141, we found that there is regression after stockfish-nnue-2020-08-30. We will investigate it.

**Comment by nodchip** [comment]
I trained nets with several commits, and found there is regression between https://github.com/nodchip/Stockfish/commit/58863c32436c22ea05121e039850253510d923d1 and https://github.com/nodchip/Stockfish/commit/1482e5215afa1b457418d45805bb57a25f4529f4.
![image 2020-09-14-19-37-35](https://user-images.githubusercontent.com/685625/93076474-5ba74a80-f6c2-11ea-9f44-fdbaf58f60b9.png)

**Comment by nodchip** [comment]
noobpwnftw independentry investigated this issue, and found there is a bug inhttps://github.com/nodchip/Stockfish/blob/ea5d437dbbebb10662f51c49f9cb36c6f100835a/src/learn/learn.cpp#L852.
https://discordapp.com/channels/435943710472011776/737073080081317948/755002405573099550

 I will fix it locally, train a net file, and compare it with 2020-07-19's one on fishtest.

**Comment by nodchip** [comment]
To confirm if the change fixes the issue, I followed the steps below:

1. Created training data.
2. Created validation data.
3. Trained a net with the change.
4. Compared 3. and the net trained with https://github.com/nodchip/Stockfish/releases/tag/stockfish-nnue-2020-07-19 .

At first, I created training data with https://github.com/nodchip/Stockfish/commit/bcfe28b2ae468d045e2f96b659401422531bacba and the following commands.
```
uci
setoption name Use NNUE value false
setoption name Threads value 27
setoption name Hash value 32768
setoption name SyzygyPath value D:\data\chess\tablebase\3-4-5
isready
gensfen depth 6 loop 100000000 use_draw_in_training_data_generation 1 eval_limit 32000 output_file_name F:\stockfish\training_data\2020-09-12\Threads=27.Hash=32768.depth=6.loop=100000000.use_draw_in_training_data_generation=1.eval_limit=32000.bin
quit
``` 

Then, I created validation data with https://github.com/nodchip/Stockfish/commit/bcfe28b2ae468d045e2f96b659401422531bacba and the following commands.
```
uci
setoption name Use NNUE value false
setoption name Threads value 27
setoption name Hash value 32768
setoption name SyzygyPath value D:\data\chess\tablebase\3-4-5
isready
gensfen depth 6 loop 1000000 use_draw_in_training_data_generation 1 eval_limit 32000 output_file_name F:\stockfish\validation_data\2020-09-12\generated_kifu.bin
quit
```

Then, I trained nets with the following commits and options to investigate this issue.

1. https://github.com/nodchip/Stockfish/commit/36092b855a5e2bcfb587a36e2055ea068e4bd8e5 (=https://github.com/nodchip/Stockfish/releases/tag/stockfish-nnue-2020-07-19)
```
uci
setoption name SkipLoadingEval value true
setoption name Threads value 16
setoption name EvalSaveDir value F:\stockfish\eval\2020-07-19.36092b855a5e2bcfb587a36e2055ea068e4bd8e5
isready
learn targetdir F:\stockfish\training_data\2020-09-12 loop 100 batchsize 1000000 use_draw_in_training 1 use_draw_in_validation 1 eta 1 lambda 1 eval_limit 32000 nn_batch_size 1000 newbob_decay 0.5 eval_save_interval 100000000 loss_output_interval 1000000 mirror_percentage 0 validation_set_file_name F:\stockfish\validation_data\2020-09-12\generated_kifu.bin
quit
```

2. https://github.com/nodchip/Stockfish/commit/35f04aaf24d05ec7b71c2cc1710d00642aaf35ec (=https://github.com/nodchip/Stockfish/releases/tag/stockfish-nnue-2020-08-11)
```
uci
setoption name SkipLoadingEval value true
setoption name Use NNUE value true
setoption name Threads value 16
setoption name EvalSaveDir value F:\stockfish\eval\2020-08-11.35f04aaf24d05ec7b71c2cc1710d00642aaf35ec
isready
learn targetdir F:\stockfish\training_data\2020-09-12 loop 100 batchsize 1000000 use_draw_in_training 1 use_draw_in_validation 1 eta 1 lambda 1 eval_limit 32000 nn_batch_size 1000 newbob_decay 0.5 eval_save_interval 100000000 loss_output_interval 1000000 mirror_percentage 0 validation_set_file_name F:\stockfish\validation_data\2020-09-12\generated_kifu.bin
quit
```

3. https://github.com/nodchip/Stockfish/tree/get_shallow_value-2020-09-14
```
uci
setoption name SkipLoadingEval value true
setoption name Use NNUE value pure
setoption name Threads value 8
setoption name EvalSaveDir value F:\stockfish\eval\2020-09-14.get_shallow_value-2020-09-14
isready
learn targetdir F:\stockfish\training_data\2020-09-12 loop 100 batchsize 1000000 use_draw_in_training 1 use_draw_in_validation 1 eta 1 lambda 1 eval_limit 32000 nn_batch_size 1000 newbob_decay 0.5 eval_save_interval 100000000 loss_output_interval 1000000 mirror_percentage 0 validation_set_file_name F:\stockfish\validation_data\2020-09-12\generated_kifu.bin
quit
```

The results of the trainings is below.

![image 2020-09-15-23-06-54](https://user-images.githubusercontent.com/685625/93225177-e5334700-f7ac-11ea-9b4d-ef7bff96f9d3.png)

The test cross entropy and the move accuracy of get_shallow_value-2020-09-14 is between 2020-07-19 and 2020-08-11.  It shows get_shallow_value-2020-09-14 comes closer to 2020-07-19.  But it is not identical to 2020-07-19.  I'm not sure if this is because another regression or randomness.

I submitted a test to fishtest (https://tests.stockfishchess.org/tests/view/5f60d07c8fbc1c8a3f476f7c).  I will merge the branch into the master if the result is better than 2020-08-30 vs 2020-07-19 (https://tests.stockfishchess.org/tests/view/5f5e0d7e61bdd7a0007e76fc).

**Comment by Sopel97** [comment]
hmm. The fishtest results are quite surprising. If I read them right there was no change in playing strength?

**Comment by nodchip** [comment]
I agree with you.  I will create a pull request, and merge it.

According to https://github.com/nodchip/Stockfish/issues/141#issuecomment-691913322, there is a -6.79 elo regression between https://github.com/nodchip/Stockfish/commit/36092b855a5e2bcfb587a36e2055ea068e4bd8e5 (2020-07-19) and https://github.com/nodchip/Stockfish/commit/35f04aaf24d05ec7b71c2cc1710d00642aaf35ec (2020-08-11), and a -19.52 elo regression between https://github.com/nodchip/Stockfish/commit/35f04aaf24d05ec7b71c2cc1710d00642aaf35ec (2020-08-11) and https://github.com/nodchip/Stockfish/commit/bc90567e0933a49f8c1911eba45ec6743a37c1ec (2020-08-30).  Could we continue the investigation in https://github.com/nodchip/Stockfish/issues/151?

**Comment by vondele** [comment]
@nodchip, I would have a question on how reproducible the results of a training run on the same data set is. For example, using the same binary, input data, and training parameters, but initializing the net randomly with different seeds, the nets will be different, I assume. Have we ever measured he spread in Elo performance of such nets? That would be an interesting data point to have as well.

**Comment by nodchip** [comment]
> Have we ever measured he spread in Elo performance of such nets?

Not yet.  I think that we should measure it.

I experimented with setting the random seeds to the current tick.  It decreased reproductivity, and we should avoid it.  I will set the random seed to a constant value next time.

**Comment by vondele** [comment]
Would be great if it could be measured. I think that will be the resolution of such tests, roughly. 

While it might help to use the same seed, it will not consistently yield nets of similar performance. Probably it often helps, but from time to time the optimization 'dynamics' will jump between different global minima nevertheless.

Thanks BTW for doing these tests in a structured way and posting the commands. That will be useful for people to reproduce.

**Comment by nodchip** [comment]
I found a mistake in https://tests.stockfishchess.org/tests/view/5f60d07c8fbc1c8a3f476f7c.  The branch was created from the nodchip's master, not the official's master.  I recreated the branch to measure the Elo, and submitted a new test. https://tests.stockfishchess.org/tests/view/5f62bac8912c15f19854b88e

I will reopen this issue.

**Comment by nodchip** [comment]
https://github.com/nodchip/Stockfish/commit/bc90567e0933a49f8c1911eba45ec6743a37c1ec (2020-08-30) vs https://github.com/nodchip/Stockfish/commit/36092b855a5e2bcfb587a36e2055ea068e4bd8e5 (2020-07-19)
https://tests.stockfishchess.org/tests/view/5f5e0d7e61bdd7a0007e76fc
ELO: -28.50 +-1.8 (95%) LOS: 0.0%
Total: 60000 W: 9834 L: 14744 D: 35422
Ptnml(0-2): 1718, 8785, 12979, 5725, 793

https://github.com/nodchip/Stockfish/commit/3260dded1e44c8fdf888de33ccd4e5816be0ede9 (get_shallow_depth) vs https://github.com/nodchip/Stockfish/commit/36092b855a5e2bcfb587a36e2055ea068e4bd8e5 (2020-07-19)
https://tests.stockfishchess.org/tests/view/5f62bac8912c15f19854b88e
ELO: -22.09 +-1.7 (95%) LOS: 0.0%
Total: 60000 W: 9821 L: 13631 D: 36548
Ptnml(0-2): 1489, 8265, 13575, 5909, 762

The test shows that get_shallow_depth() is fixed.  And the result is better than https://github.com/nodchip/Stockfish/commit/bc90567e0933a49f8c1911eba45ec6743a37c1ec (2020-08-30) vs https://github.com/nodchip/Stockfish/commit/36092b855a5e2bcfb587a36e2055ea068e4bd8e5 (2020-07-19).  I will close this issue.
